### PR TITLE
flux: based helm install of kepler

### LINKED ISSUE #15: [Automated] installing kepler using flux
## Description
it is going to use flux specific kubernetes resources which will be backed by GitOps.

## Progress
- [x] Find docs as a reference to install
- [x] Create the flux manifest files

## Issue
- Parent: #5 

cc @rossf7

### PR DESCRIPTION
**Pull request by dipankardas011**
## Solution
References
- https://fluxcd.io/flux/components/helm/helmreleases/#install-configuration
- https://artifacthub.io/packages/helm/kepler/kepler

we need `--create-namsepace` flag to be true according to the docs

## Issue 
- Closes #15

## Note for Reviewer
Do remove the `clusters/kube-prometheus.yaml` as its been worked on by #12

### CODE REVIEW

File: clusters/kepler.yaml [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  namespace: kepler  # TODO: fix some better place where to store
+spec:
+  interval: 2m
+  url: https://sustainable-computing-io.github.io/kepler-helm-chart
+---
+apiVersion: helm.toolkit.fluxcd.io/v2beta2
+kind: HelmRelease
+metadata:
+  name: kepler
+  namespace: kepler
```
Comment (rossf7) [inline clusters/kepler.yaml:16]:
```suggestion
  namespace: flux-system
```

File: clusters/kepler.yaml [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+spec:
+  interval: 5m
+  # https://fluxcd.io/flux/components/helm/helmreleases/#install-configuration
+  install:
+    createNamespace: true
+  chart:
+    spec:
+      chart: kepler/kepler
```
Comment (rossf7) [inline clusters/kepler.yaml:24]:
```suggestion
      chart: kepler
```

Chart should just be `kepler` the prefix is used by the helm CLI to reference the repo.

File: clusters/kepler.yaml [hunk 3]
```diff
@@ -0,0 +1,30 @@
+# Refer: https://artifacthub.io/packages/helm/kepler/kepler?modal=install
+---
+apiVersion: source.toolkit.fluxcd.io/v1beta2
+kind: HelmRepository
+metadata:
+  name: kepler-repo
+  namespace: kepler  # TODO: fix some better place where to store
```
[the remark this answers is not in the document]
Comment (rossf7) [reply clusters/kepler.yaml:7]: @dipankardas011 Can we use these settings to match what @nikimanoledaki used here? 
https://github.com/nikimanoledaki/sustainability-journey-with-gitops/blob/main/clusters/kepler.yaml

File: clusters/kepler.yaml [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+apiVersion: helm.toolkit.fluxcd.io/v2beta2
+kind: HelmRelease
+metadata:
+  name: kepler
+  namespace: kepler
+spec:
+  interval: 5m
+  # https://fluxcd.io/flux/components/helm/helmreleases/#install-configuration
+  install:
+    createNamespace: true
```
Comment (rossf7) [inline clusters/kepler.yaml:21]: Good idea to use create namespace here 👍

File: clusters/kepler.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (dipankardas011) [reply clusters/kepler.yaml:7]: have to check what all changes does helm and kustomize have for kepler

File: clusters/kepler.yaml (hunk 1 above)
Comment (dipankardas011) [reply clusters/kepler.yaml:16]: but the docs mention that it is installed under namespace kepler

Comment (dipankardas011) [reply clusters/kepler.yaml:16]: We can modify it refering to this https://sustainable-computing.io/installation/kepler-helm/
but do we want to?

**Comment by dipankardas011** [comment]
@rossf7 the latest Commit makes the flux install kepler in the monitoring namespace due to one issue
Refer: [Update-Prometheus](https://sustainable-computing.io/installation/kepler-operator/#scrape-all-namespaces), [For the service monitor to work](https://sustainable-computing.io/installation/kepler-operator/)

**Comment by dipankardas011** [comment]
To test out
```bash
kubectl apply -f clusters/kube-prometheus.yaml
kubectl apply -f clusters/kepler.yaml

kubectl -n monitoring port-forward svc/grafana 3000:3000
```
to get the dashboard create a new one and paste this https://raw.githubusercontent.com/sustainable-computing-io/kepler/main/grafana-dashboards/Kepler-Exporter.json

to you get this 
image

**Comment by dipankardas011** [comment]
> @rossf7 the latest Commit makes the flux install kepler in the monitoring namespace due to one issue Refer: [Update-Prometheus](https://sustainable-computing.io/installation/kepler-operator/#scrape-all-namespaces), [For the service monitor to work](https://sustainable-computing.io/installation/kepler-operator/)

May be for now we can try with 
```
services                     []                 []              [get list watch]
endpoints                    []                 []              [get list watch]
```

File: clusters/kube-prometheus.yaml [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+              - ""
+              resources:
+              - pods
+              verbs:
+              - get
+              - list
+              - watch
+      target:
+        kind: ClusterRole
+        name: prometheus-k8s
```
Comment (dipankardas011) [inline clusters/kube-prometheus.yaml:91]: Fix for serviceMonitor not found

File: clusters/kepler.yaml [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+  namespace: flux-system
+# https://fluxcd.io/flux/components/helm/helmreleases/#install-configuration
+spec:
+  interval: 24h
+  # https://fluxcd.io/flux/components/helm/helmreleases/#inline-values
+  values:
+    serviceMonitor:
+      enabled: "true"
+      namespace: "monitoring"
+  targetNamespace: kepler # Note: By default kube-prometheus does not let you scrape services deployed in namespaces other than monitoring. So if you are running Kepler outside monitoring
```
Comment (dipankardas011) [inline clusters/kepler.yaml:25]: kept this comment to remind us that kube-promtheus needs to be patched

**Review by rossf7 — APPROVED** [review]
LGTM

@dipankardas011 Kepler changes look good and thanks for finding the solution to the scraping problem. 

I think we do want Prometheus to be able to scrape services in all namespaces.

**Comment by dipankardas011** [comment]
> LGTM
> 
> @dipankardas011 Kepler changes look good and thanks for finding the solution to the scraping problem. 
> 
> I think we do want Prometheus to be able to scrape services in all namespaces.

I was wondering whether we should keep the kube-prometeus yaml which I created as issue #12 is working on that

**Comment by rossf7** [comment]
> I was wondering whether we should keep the kube-prometeus yaml which I created as issue https://github.com/cncf-tags/green-reviews-tooling/pull/12 is working on that

@dipankardas011 Yes I think it might be easier to remove kube-prometheus.yaml and patch the cluster role in #12 

@AntonioDiTuri Is that OK with you?

cc @nikimanoledaki

**Review by dipankardas011 — COMMENTED** [review]
@rossf7 completed refactoring the PR you can now merge it

File: clusters/kepler.yaml [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+kind: HelmRepository
+metadata:
+  name: kepler
+  namespace: flux-system
+spec:
+  interval: 24h
+  url: https://sustainable-computing-io.github.io/kepler-helm-chart
+---
+# https://fluxcd.io/flux/components/helm/helmreleases/#install-configuration
+apiVersion: helm.toolkit.fluxcd.io/v2beta2
```
Comment (rossf7) [inline clusters/kepler.yaml:15]:
```suggestion
apiVersion: helm.toolkit.fluxcd.io/v2beta1
```

Cluster is running Flux v2.1.2 so we need to use v2beta1 until we upgrade.

File: clusters/kepler.yaml [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+  install:
+    createNamespace: true
+  chart:
+    spec:
+      chart: kepler
+      version: '0.5.3'
+      sourceRef:
+        kind: HelmRepository
+        name: kepler
+...
```
Comment (nikimanoledaki) [inline clusters/kepler.yaml:37]:
```suggestion
```

**Comment by dipankardas011** [comment]
Okay let me rebase it

**Comment by dipankardas011** [comment]
done 🚀

**Review by nikimanoledaki — APPROVED** [review]
Let's ship it!!! Thank you @dipankardas011 @rossf7! 🥳

**Pull request merged**